src/adaptors/tcp/tcp_adaptor.c

```diff
@@ -474,10 +495,24 @@ static int handle_incoming(qdr_tcp_connection_t *conn, const char *msg)
         // Add the application properties
         //
         props = qd_compose(QD_PERFORMATIVE_APPLICATION_PROPERTIES, props);
-        qd_compose_start_map(props);
+        qd_compose_start_map(props);  // start map
         qd_compose_insert_symbol(props, QD_AP_FLOW_ID);
         vflow_serialize_identity(conn->vflow, props);
-        qd_compose_end_map(props);
+        if (conn->require_tls) {
```

Comment:
Can't the negotiated ALPN be retrieved once during the "on_secure" callback (storing it in the tcp connection structure)?  Then you can simply check if "conn->negotiated_alpn" (or whatever name) is set and then insert it in the map? This function is pretty long already.

Reply:
Done !